Adding documentation for `qml.breakpoint()` and `qml.PLDB` (#5789)
PennyLaneAI/pennylane: doc/code/qml_debugging.rst

Reviewer: In addition to this module, it would be nice to have a short section just below [here](https://xanaduai-pennylane--5789.com.readthedocs.build/en/5789/introduction/inspecting_circuits.html#debugging-with-mid-circuit-snapshots) in the quickstart.

Author: Added 👍🏼

by provding access to the internal state of the circuit and the ``QuantumTape`` as 
the circuit operations are applied. The functionality is highlighted by debugging 
and example circuit.

Consider the following circuit, 

.. code-block:: python

    import pennylane as qml
    
    @qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
    def circuit(x):   
        # 1) Prepare equal probabilities over wires (0,1):
        qml.breakpoint()
        qml.Hadamard(wires=0)
        
        # 2) Entangle wires (1,2):
        qml.CNOT(wires=(0,2))

        # 3) Apply a rotation of angle 2x to each wire:
        for w in (0, 1, 2):
            qml.RX(2*x, wires=w)

        # 4) Reset wire 1: 
        qml.breakpoint()
        qml.RX(-x, wires=1)

        return qml.sample()

    circuit(1.2345)

While executing this circuit once a `qml.breakpoint()` is encountered, an interactive
debugging prompt is launched. The prompt specifies the path to the script along with
the next line to be executed after the breakpoint.

.. code-block:: console

    > /Users/your/path/to/script.py(7)circuit()
    -> qml.Hadamard(wires=0)
    [pldb]:

> We can interact with the prompt using the builtin commands: ``list`` , ``next``, 

Author: They are builtins from Pdb so we don't have explicit docs for them

qml.debugging
=============

.. automodapi:: pennylane.debugging
    :no-heading:
    :no-inherited-members:
    :skip: PLDB

>     :skip: pldb_device_manager

Reviewer: Is it intended for snapshots to show up here? :thinking: 
Overall, it might make sense. Just curious.

Reviewer: Though `snapshots` is also a top-level accessible function but it's showing up as `qml.debugging.snapshots`

Author: It was intended on purpose to place snapshots here. Originally, the snapshots logic and the debugger were both located in the same python file.

Entering the Debugging Context
------------------------------

> The :func:`~.breakpoint <qml.breakpoint()>` function provides an interface for interacting with and

Reviewer 2: The link is broken.

qml.debugging
=============

.. automodapi:: pennylane.debugging
    :no-heading:
    :no-inherited-members:
    :skip: _Debugger
    :skip: contextmanager
    :skip: _measure
    :skip: PLDB
    :skip: pldb_device_manager

Entering the Debugging Context
------------------------------

> The ``qml.breakpoint()`` function provides an interface for interacting with and

Reviewer 3: ```suggestion
The :func:`~.breakpoint <qml.breakpoint()>`` function provides an interface for interacting with and
```

stepping through a quantum circuit during execution. It allows for faster debugging

> by provding access to the internal state of the circuit and the ``QuantumTape`` as 

Author: I spoke with Christina, to users we call it QuantumTape, but internally the object is a QuantumScript.

---------------------------------------------------

The Pennylane Debugger (pldb) is built from the native python debugger (pdb), as such 
it shares a similar interface. We can interact with the debugger using the 
builtin commands: ``list``, ``longlist``, ``next``, ``continue``, and ``quit``. Any 
variables defined in the scope of the quantum function can also be accessed from the 
debugger.

.. code-block:: console

    [pldb]: print(x)
    1.2345

The ``list`` (and ``longlist``) command will print a section of code around the 
breakpoint, highlighting the next line to be executed. This can be used to determine
the location of execution in the circuit.

.. code-block:: console

    [pldb]: longlist
      3  	@qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
      4  	def circuit(x):
      5  	    qml.breakpoint()
      6  	
      7  ->	    qml.Hadamard(wires=0)
      8  	    qml.CNOT(wires=(0,2))
      9  	
     10  	    for w in (0, 1, 2):
     11  	        qml.RX(2*x, wires=w)
     12  	
     13  	    qml.breakpoint()
     14  	    qml.RX(-x, wires=1)
     15  	    return qml.sample()
    
The ``next`` command will execute the next line of code, and print the following 

> line to be executed (eg. the next operation to execute is ``CNOT``).

Reviewer 2: ```suggestion
line to be executed, e.g., the next operation to execute is ``CNOT``.
```

.. code-block:: console
    
    [pldb]: next
    > /Users/your/path/to/script.py(8)circuit()
    -> qml.CNOT(wires=(0,2))
    [pldb]: list
      3  	@qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
      4  	def circuit(x):
      5  	    qml.breakpoint()
      6  	
      7  	    qml.Hadamard(wires=0)
      8  ->	    qml.CNOT(wires=(0,2))
      9  	
     10  	    for w in (0, 1, 2):
     11  	        qml.RX(2*x, wires=w)
     12  	
     13  	    qml.breakpoint()

> Alternative, ``continue`` allows for jumping between breakpoints. This command resumes

Reviewer 2: The ``continue`` command ...

Author: Fixed

> Alternative, ``continue`` allows for jumping between breakpoints. This command resumes

Author: ```suggestion
Alternative, the ``continue`` command allows for jumping between breakpoints. This command resumes
```
